File: B형대비/추가+문제/[실전실습코드] 이승재_블록쌓기게임.java
import java.io.*;
import java.util.*;
class Solution {
    private static BufferedReader br;
    private static UserSolution usersolution = new UserSolution();

    private final static int CMD_INIT = 100;
    private final static int CMD_DROP = 200;
	
	
	public static final class Result{
		int top;
		int count;

		Result(){
			top = 0;
			count = 0;
		}
	}

    private static boolean run() throws Exception {

        StringTokenizer stdin = new StringTokenizer(br.readLine(), " ");

        int query_num = Integer.parseInt(stdin.nextToken());
		int ans_top, ans_count;
        boolean ok = false;

        for (int q = 0; q < query_num; q++) {
            stdin = new StringTokenizer(br.readLine(), " ");
            int query = Integer.parseInt(stdin.nextToken());

            if (query == CMD_INIT) {
                int C = Integer.parseInt(stdin.nextToken());
                usersolution.init(C);
                ok = true;
            } else if (query == CMD_DROP) {
                int mCol = Integer.parseInt(stdin.nextToken());
                int mHeight = Integer.parseInt(stdin.nextToken());
                int mLength = Integer.parseInt(stdin.nextToken());
				
				Result ret = usersolution.dropBlocks(mCol, mHeight, mLength);
                ans_top = Integer.parseInt(stdin.nextToken());
				ans_count = Integer.parseInt(stdin.nextToken());
                
                if (ans_top != ret.top || ans_count != ret.count) {
                    ok = false;
                }
            }
        }
        return ok;
    }

    public static void main(String[] args) throws Exception {
        int T, MARK;

        // System.setIn(new java.io.FileInputStream("res/sample_input.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stinit = new StringTokenizer(br.readLine(), " ");
        T = Integer.parseInt(stinit.nextToken());
        MARK = Integer.parseInt(stinit.nextToken());

        for (int tc = 1; tc <= T; tc++) {
            int score = run() ? MARK : 0;
            System.out.println("#" + tc + " " + score);
        }

        br.close();
    }
}

class UserSolution {
	int[] a;
	int[] m;
	int[] M;
	long tot;
	int col;
    void init(int C) {
		int p2 = 1;
		while (p2 < C) p2 *= 2;
		a = new int[p2 * 2];
		m = new int[p2 * 2];
		M = new int[p2 * 2];
		tot = 0;
		col = C;
    }
	void range_update(int qs, int qe, int qk, int ns, int ne, int nx) {
		if (qe < ns || ne < qs) return;
		if (qs <= ns && ne <= qe) {
			a[nx] += qk; m[nx] += qk; M[nx] += qk;
		} else {
			int mid = (ns + ne) / 2;
			range_update(qs, qe, qk, ns, mid, nx * 2);
			range_update(qs, qe, qk, mid + 1, ne, nx * 2 + 1);
			m[nx] = Math.min(m[nx * 2], m[nx * 2 + 1]) + a[nx];
			M[nx] = Math.max(M[nx * 2], M[nx * 2 + 1]) + a[nx];
		}
	}

    Solution.Result dropBlocks(int c, int height, int length) {
        Solution.Result ret = new Solution.Result();
		tot += height * length;
        range_update(c, c + length - 1, height, 0, col - 1, 1);
		int mn = m[1], mx = M[1];
		ret.top = mx - mn;
		ret.count = (int)((tot - 1L * mn * col) % 1000000);
		return ret;
    }
}

File: B형대비/추가+문제/[실전실습코드] 이승재_블록쌓기게임.cpp
#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS
#endif

#include <stdio.h>

struct Result
{
    int top;
    int count;
};

void init(int C);
Result dropBlocks(int mCol, int mHeight, int mLength);

#define CMD_INIT 100
#define CMD_DROP 200

static bool run()
{
    int query_num;
    scanf("%d", &query_num);

    int ans_top, ans_count;
    bool ok = false;

    for (int q = 0; q < query_num; q++)
    {
        int query;
        scanf("%d", &query);
        if (query == CMD_INIT)
        {
            int C;
            scanf("%d", &C);
            init(C);
            ok = true;
        }
        else if (query == CMD_DROP)
        {
            int mCol, mHeight, mLength;
            scanf("%d %d %d", &mCol, &mHeight, &mLength);
            Result ret = dropBlocks(mCol, mHeight, mLength);
            scanf("%d %d", &ans_top, &ans_count);
            if (ans_top != ret.top || ans_count != ret.count)
            {
                ok = false;
            }
        }
    }
    return ok;
}

int main()
{
    setbuf(stdout, NULL);
    // freopen("sample_input.txt", "r", stdin);
    int T, MARK;
    scanf("%d %d", &T, &MARK);

    for (int tc = 1; tc <= T; tc++)
    {
        int score = run() ? MARK : 0;
        printf("#%d %d\n", tc, score);
    }
    return 0;
}

#include <bits/stdc++.h>
using namespace std;
 
constexpr int base = 1 << 20;
int add[base << 1];
int min_value[base << 1];
int max_value[base << 1];
int col;
template<int qk> void range_update(int qs, int qe, int ns = 0, int ne = -1, int nx = 1) {
    if (ne == -1) ne = col - 1;
    if (qe < ns || ne < qs) return;
    if (qs <= ns && ne <= qe) {
        add[nx] += qk;
		min_value[nx] += qk;
		max_value[nx] += qk;
    } else {
        int mid = (ns + ne) / 2;
        range_update<qk>(qs, qe, ns, mid, nx * 2);
        range_update<qk>(qs, qe, mid + 1, ne, nx * 2 + 1);
        min_value[nx] = min(min_value[nx * 2], min_value[nx * 2 + 1]) + add[nx];
        max_value[nx] = max(max_value[nx * 2], max_value[nx * 2 + 1]) + add[nx];
    }
}
long long tot = 0;
void init(int C) {
    tot = 0;
    memset(add, 0, sizeof add);
    memset(min_value, 0, sizeof min_value);
    memset(max_value, 0, sizeof max_value);
    col = C;
}
Result dropBlocks(int mCol, int mHeight, int mLength) {
    Result ret;
    ret.top = 0;
    ret.count = 0;
    tot += mHeight * mLength;

    if (mHeight == 1) range_update<1>(mCol, mCol + mLength - 1);
    if (mHeight == 2) range_update<2>(mCol, mCol + mLength - 1);
    if (mHeight == 3) range_update<3>(mCol, mCol + mLength - 1);
    int mn = min_value[1];
    int mx = max_value[1];
    ret.top = mx - mn;
    ret.count = (tot - 1LL * mn * col) % 1'000'000;
    return ret;
}